lib/IDE/CursorInfo.cpp

```diff
@@ -0,0 +1,278 @@
+//===--- CursorInfo.cpp ---------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/IDE/CursorInfo.h"
+#include "ExprContextAnalysis.h"
+#include "swift/AST/ASTDemangler.h"
+#include "swift/AST/GenericEnvironment.h"
+#include "swift/AST/NameLookup.h"
+#include "swift/AST/USRGeneration.h"
+#include "swift/IDE/TypeCheckCompletionCallback.h"
+#include "swift/Parse/CodeCompletionCallbacks.h"
+#include "swift/Sema/ConstraintSystem.h"
+#include "swift/Sema/IDETypeChecking.h"
+#include "clang/AST/Attr.h"
+#include "clang/AST/Decl.h"
+#include "clang/Basic/Module.h"
+
+using namespace swift;
+using namespace swift::constraints;
+using namespace ide;
+
+namespace {
+
+// MARK: - Utilities
+
+void typeCheckDeclAndParentClosures(ValueDecl *VD) {
+  // We need to type check any parent closures because their types are
+  // encoded in the USR of ParentContexts in the cursor info response.
+  auto DC = VD->getDeclContext();
+  while (DC->getParent()) {
+    if (auto Closure = dyn_cast<AbstractClosureExpr>(DC)) {
+      if (Closure->getType().isNull()) {
+        typeCheckASTNodeAtLoc(
+            TypeCheckASTNodeAtLocContext::declContext(DC->getParent()),
+            Closure->getLoc());
+      }
+    }
+    DC = DC->getParent();
+  }
+
+  typeCheckASTNodeAtLoc(
+      TypeCheckASTNodeAtLocContext::declContext(VD->getDeclContext()),
+      VD->getLoc());
+}
+
+// MARK: - NodeFinderResults
+
+enum class NodeFinderResultKind { Decl };
+
+class NodeFinderResult {
+  NodeFinderResultKind Kind;
+
+protected:
+  NodeFinderResult(NodeFinderResultKind Kind) : Kind(Kind) {}
+
+public:
+  NodeFinderResultKind getKind() const { return Kind; }
+};
+
+class NodeFinderDeclResult : public NodeFinderResult {
+  ValueDecl *ValueD;
+
+public:
+  NodeFinderDeclResult(ValueDecl *ValueD)
+      : NodeFinderResult(NodeFinderResultKind::Decl), ValueD(ValueD) {}
+
+  ValueDecl *getDecl() const { return ValueD; }
+
+  static bool classof(const NodeFinderResult *Res) {
+    return Res->getKind() == NodeFinderResultKind::Decl;
+  }
+};
+
+// MARK: - NodeFinder
+
+/// Walks the AST, looking for a node at \c LocToResolves. While walking the
+/// AST, also gathers information about shorthand shadows.
+class NodeFinder : ASTWalker {
```

Comment:
I’ll see what I can do in a follow-up PR.

Reply:
Yeah, IMO ideally we'd do what Clang does here - ie. build up a hierarchy to the deepest node and pass that around everywhere we need it (rather than continually performing the same ASTWalk skipping outside-of-loc ranges in basically every refactoring + cursor info).

Reply:
I filed rdar://103178329 so we don’t forget about it and hope that I’ll get to it at some point.